airbyte-ci/connectors/metadata_service/lib/metadata_service/gcs_upload.py

```diff
@@ -86,7 +90,46 @@ def upload_file_if_changed(
     return False, remote_blob.id
 
 
-def upload_metadata_to_gcs(bucket_name: str, metadata_file_path: Path) -> Tuple[bool, str]:
+def _latest_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket.Bucket, metadata_file_path: Path) -> Tuple[bool, str]:
+    version_path = get_metadata_remote_file_path(metadata.data.dockerRepository, metadata.data.dockerImageTag)
+    latest_path = get_metadata_remote_file_path(metadata.data.dockerRepository, "latest")
+    latest_icon_path = get_icon_remote_file_path(metadata.data.dockerRepository, "latest")
+
+    (
+        version_uploaded,
+        version_blob_id,
+    ) = upload_file_if_changed(metadata_file_path, bucket, version_path)
+    latest_uploaded, _latest_blob_id = upload_file_if_changed(metadata_file_path, bucket, latest_path)
+
+    # Replace metadata file name with icon file name
+    local_icon_path = metadata_file_path.parent / ICON_FILE_NAME
+    if local_icon_path.exists():
+        upload_file_if_changed(local_icon_path, bucket, latest_icon_path)
+
+    return version_uploaded or latest_uploaded, version_blob_id
+
+
+def _prerelease_upload(metadata: ConnectorMetadataDefinitionV0, bucket: storage.bucket.Bucket, prerelease_tag: str) -> Tuple[bool, str]:
+    # replace any dockerImageTag references with the actual tag
+    # this includes metadata.data.dockerImageTag, metadata.data.registries[].dockerImageTag
+    # where registries is a dictionary of registry name to registry object
+    metadata_dict = to_json_sanitized_dict(metadata, exclude_none=True)
+    metadata_dict["data"]["dockerImageTag"] = prerelease_tag
+    for registry in get(metadata_dict, "data.registries", {}).values():
+        if "dockerImageTag" in registry:
+            registry["dockerImageTag"] = prerelease_tag
```

Comment:
This feels a bit gnarly - I'm not sure about the overrides part?

Not that they matter because this isn't going into a registry anyway, but I don't think publishing a pre-release inherently makes any changes to the overrides. But you might be considering only the no overrides case and I might be considering only the overrides case

`metadata_dict["data"]["dockerImageTag"]` will definitely have to be up to date when we grab the metadata from the platform and persist the definition based off of it, if we go that route, so good catch there!

Reply:
This is the tricky part!

So for example if you are the platform and say you are using the LD flag to override the connector image to `airbyte/source-postgres:1.1.1-dev.123asd` you would go looking for that info at either

https://connectors.airbyte.com/files/metadata/airbyte/source-postgres/1.1.1-dev.123asd/oss.json

or

https://connectors.airbyte.com/files/metadata/airbyte/source-postgres/1.1.1-dev.123asd/cloud.json

Now if the original metadata had `registries.cloud.dockerImageTag` pinned to 1.1.0
and we didnt overwrite `registries.cloud.dockerImageTag` with the prerelease tag
then the the cloud use would still get `airbyte/source-postgres:1.1.1`
because thats what the registry entry would have

Reply:
ahaaaa

that makes sense! 

Maybe a comment as to "if we request the registry entry for a prerelease version specifically, we want to make sure we do not get a different overridden version" or something